### PR TITLE
Feat game config

### PR DESCRIPTION
**Pull request by parkerpersaud**
https://trello.com/c/FFUT5aqJ/50-425-game-configuration-menu

### CODE REVIEW

File: src/tictactoe/game_runner.clj [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+(def game-types
+  {:1 "Human vs. Human" :2 "Human vs. Easy Computer"})
+
+(def board-sizes
+  {:1 "3x3" :2 "4x4"})
+
+
+(defn start []
+  (let [game-options (game-menu/menu game-types board-sizes)]
```
Comment (andarcabrera) [inline src/tictactoe/game_runner.clj]: is the `let` necessary here? can you call the function directly?

File: src/tictactoe/game_menu.clj [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+(defn pick-game [game-types] 
+  (ui/print-message (messages/game-type))
+  (ui/loop-and-print game-types)
+  (-> 
+    (ui/get-input)
+    (keyword)
+    (validator/input-validation-loop game-types)
+    (keyword)))
+
+(defn pick-board-size [board-sizes]
```
Comment (andarcabrera) [inline src/tictactoe/game_menu.clj]: `pick-game` and `pick-board-size` functions seem identical except the parameter that's passed in and the message. Can that be abstracted?

File: src/tictactoe/game_menu.clj [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    (keyword)))
+
+(defn pick-player-one-marker []
+  (ui/print-message (messages/player-one-marker))
+  (-> 
+    (ui/get-input)
+    (marker-validator/marker-validation-loop)
+    (string/upper-case)))
+
+(defn pick-player-two-marker [player1-marker]
```
Comment (andarcabrera) [inline src/tictactoe/game_menu.clj]: same for the `pick-player-marker`. Can you use an optional argument and abstract this function?

**Review by andarcabrera — COMMENTED** [review]
It's not in the diff, but you can delete the core test that was part of the original setup.

File: project.clj [hunk 4]
```diff
@@ -6,4 +6,5 @@
   :dependencies [[org.clojure/clojure "1.9.0"]]
   :main ^:skip-aot tictactoe.core
   :target-path "target/%s"
+  :plugins [[lein-cloverage "1.0.10"] [lein-eftest "0.4.3"]]
```
Comment (aaronlahey) [inline project.clj]: 👍

File: test/tictactoe/game_menu_test.clj [hunk 5]
```diff
@@ -0,0 +1,54 @@
+(ns tictactoe.game-menu-test
+  (:require [clojure.test :refer :all]
```
Comment (aaronlahey) [inline test/tictactoe/game_menu_test.clj]: `:refer :all` is a bad practice in most cases. I basically *always* prefer `:as`.

`:as` > `:refer [a b c]` > `:refer :all` > `:use`

File: test/tictactoe/game_menu_test.clj [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      (is (= :2 (with-in-str "2" (pick board-sizes (messages/board-type))))))))
+
+(deftest get-player-one-marker
+  (testing "Get first players marker and return it"
+    (with-out-str "test" (ui/print-message (messages/player-one-marker))
+      (is (= "X" (with-in-str "X" (pick-marker)))))))
+
+(deftest get-player-two-marker
+  (testing "Get second players marker and return it"
+	 (with-out-str "test" (ui/print-message (messages/player-two-marker))
```
Comment (aaronlahey) [inline test/tictactoe/game_menu_test.clj]: Some strange formatting here. and below. If you're in vim, a lot of times I'll just select the whole file and press `=` to reformat.

Comment (aaronlahey) [reply test/tictactoe/game_menu_test.clj]: `gg` go to top
`V` to select the first line
`G` to select all lines
`=` to reindent

File: test/tictactoe/game_runner_test.clj [hunk 7]
```diff
@@ -0,0 +1,13 @@
+(ns tictactoe.game-runner-test
+  (:require [clojure.test :refer :all]
+            [tictactoe.game-runner :refer :all]
+            [tictactoe.game-menu :as game-menu]
+            [tictactoe.game-messages :as messages]
+            [tictactoe.user-interface :as ui]))
+
+(defn make-input [coll]
+  (apply str (interleave coll (repeat "\n"))))
```
Comment (aaronlahey) [inline test/tictactoe/game_runner_test.clj]: `interpose` over `interleave` here.

File: src/tictactoe/user_interface.clj [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+(defn print-message
+  ([message]
+    (println message))
+  ([message supplementing-info]
+    (println message "=>" supplementing-info)))
+
+(defn get-input []
+  (read-line))
+
+(defn loop-and-print [options]
```
Comment (aaronlahey) [inline src/tictactoe/user_interface.clj]: Can you format your functions like this...

```
(defn do-something
  []
  (println "hi"))
```

File: src/tictactoe/marker_validation.clj [hunk 9]
```diff
@@ -0,0 +1,33 @@
+(ns tictactoe.marker-validation
+	(:require [tictactoe.user-interface :as ui]
```
Comment (aaronlahey) [inline src/tictactoe/marker_validation.clj]: Formatting? I'm going to start cracking down on wat formatting. When you put your PRs up just glance over them and look for formatting issues. Ideally, when you join a team, you'll have a tool which lints the code if not auto format it for you.

File: src/tictactoe/marker_validation.clj [hunk 10]
```diff
@@ -0,0 +1,33 @@
+(ns tictactoe.marker-validation
```
Comment (aaronlahey) [inline src/tictactoe/marker_validation.clj]: I feel like there are already too many namespaces for the amount of functionality you've implemented. There isn't really a 1-1 correspondence between namespaces and classes in Java and I feel like that's where you're headed.

Let's try this. Implement everything in a single namespace for now and when it becomes painful, or responsibilities start to become muddled we'll extract namespaces.

File: test/tictactoe/user_interface_test.clj [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+(deftest get-input-test
+  (is (= "input received" (with-in-str "input received" (get-input)))))
+
+(deftest test-loop-and-print
+  (is (= "game => Human vs Human\n" (with-out-str (loop-and-print {:game "Human vs Human"})))))
+
+(deftest prints-and-loops
+  (testing "It loops and prints out key value pairs")
+  (with-out-str "test"
```
Comment (aaronlahey) [inline test/tictactoe/user_interface_test.clj]: Formatting.

**Review by aaronlahey — CHANGES_REQUESTED** [review]

File: src/tictactoe/game_runner.clj [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+(defn chosen-game-options 
+  "Takes a hash-map of all game options chosen and displays the summary"
+  [game-options]
+  (print-message (game-choices-message))
+  (loop-and-print game-options))
+
+(defn menu
+  "Displays a welcome message and prompts the user for game type, board size, and player markers.
+    It displays the summary of options and returns a hash-map of choices."
```
Comment (aaronlahey) [inline src/tictactoe/game_runner.clj]: Strange indentation here. You'll want to indent your docstrings so when you call `(doc tictactoe.game-runner/menu)` at the repl it looks correctly formatted.

That's usually like this

```
(defn add
  "Returns the sum of two number. Oh look,
  a newline!"
  [x y]
  (+ x y))

File: test/tictactoe/game_runner_test.clj [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+  (is (= "game => Human vs Human\n" (with-out-str (game-runner/loop-and-print {:game "Human vs Human"})))))
+
+(deftest prints-and-loops
+  (testing "It loops and prints out key value pairs")
+    (with-out-str "test"
+      (is (= nil (game-runner/loop-and-print {:game "Human V Human" :board "3x3"})))))
+
+(deftest loops-when-invalid-marker
+  (testing "When invalid input it loops until valid input is received"
+    (is (= "Invalid. Please pick an available option.\n"
```
Comment (aaronlahey) [inline test/tictactoe/game_runner_test.clj]: You might want to check your indentation settings. Generally *most* functions should align arguments like this:

```
(is (= 42
       (+ 21 21)))
```

Certain Clojure functions just indent two spaces without aligning. I would expect `with-*`, `def*`, etc to be that way.

File: test/tictactoe/game_runner_test.clj [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+       (with-out-str (game-runner/chosen-game-options {:game 1 :board 2 :player1 "X" :player2"O"})))))))
+
+(deftest show-game-menu
+  (testing "Gets all game options from user and returns hash of options"
+    (with-out-str "test" (with-in-str (make-input '("2" "2" "a" "b")) 
+      (game-runner/menu game-runner/game-types game-runner/board-sizes))))
+        (is (= {:game "1" :board "1" :player1 "A" :player2 "B"})))
+
+(deftest test-start 
+  (with-out-str "test" (with-in-str (make-input '("1" "2" "a" "b")) (game-runner/start))))
```
Comment (aaronlahey) [inline test/tictactoe/game_runner_test.clj]: When I'm reading these tests all I can really see is `with-out-str` and `with-in-str`. Is there a way to refactor so the tests more clearly reveal what exactly they're testing?

File: src/tictactoe/game_runner.clj [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+  (:require [clojure.string :as string]))
+
+(def game-types
+  {"1" "Human vs. Human" "2" "Human vs. Easy Computer"})
+
+(def board-sizes
+  {"1" "3x3" "2" "4x4"})
+
+(def markers 
+  (set (.split "abcdefghijklmnopqrstuvwxyz" "")))
```
Comment (aaronlahey) [inline src/tictactoe/game_runner.clj]: See if you can do this without using Java interop.

**Review by aaronlahey — CHANGES_REQUESTED** [review]

**Comment by aaronlahey** [comment]
Ugh. I submitted my review before writing my final comment:

So as the code stand now there are two namespaces:

```
tictactoe.main
tictactoe.game-runner
```

`tictactoe.main` is the entry point to your *whole* program. It invokes `tictactoe.game-runner/start`. That suggests that you have a single public function to export/test, yet I see a handful of functions with docstrings and dedicated unit tests. You're creating a tighter coupling between your production code and your unit tests than needs to exist.

Since there's only one function which is being used outside of `game-runner` (`start`) lets declare every other function as private:

```
(defn- foo
  "My least favorite way of declaring private because it's hard to see"
  []
  (+ 1 2))

(defn ^:private foo
  "Works just fine.
  []
  (+ 1 2))

(defn ^{:private true} foo
  "Verbose but has the added benefit of having a consistent syntax when multiple keys need to be added"
  []
  (+ 1 2))
```

**Comment by aaronlahey** [comment]
I also get test failures when I run `lein test` locally. Can you confirm your CI is working correctly?

**Pull request merged**